File: tools/flows.py
"""WO-013 phase 6: one registry, no open loops.

A flow that can be entered and not left is a dead end, and the review found several. Finding them
one at a time does not last. This makes a flow with no ending impossible to ship.

Run it: python3 tools/flows.py
"""
import threading, http.server, socketserver, functools, os, sys, json, subprocess, shutil

os.environ.setdefault("PLAYWRIGHT_BROWSERS_PATH", "/opt/pw-browsers")
ROOT = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
CH = "/opt/pw-browsers/chromium-1194/chrome-linux/chrome"
Handler = functools.partial(http.server.SimpleHTTPRequestHandler, directory=ROOT)
socketserver.TCPServer.allow_reuse_address = True
httpd = socketserver.TCPServer(("127.0.0.1", 0), Handler)
PORT = httpd.server_address[1]
httpd.daemon_threads = True
threading.Thread(target=httpd.serve_forever, daemon=True).start()
base = f"http://127.0.0.1:{PORT}"

from playwright.sync_api import sync_playwright

fails = []
def ck(n, c, d=None):
    print(("PASS " if c else "FAIL ") + n)
    if not c:
        fails.append(n)
        if d is not None: print("      " + str(d)[:400])

SEED = """()=>{ const now=Date.now();
 localStorage.setItem('thrive_opps_v1', JSON.stringify([
  { slug:'one', business:'One Co', published:true, up:now, contact_tier:'A',
    channel:{kind:'email', to:'a@one.example'}, outreach_text:'Hello.' }]));
}"""


def session(b, relay=None):
    ctx = b.new_context(viewport={"width": 1280, "height": 900})
    if relay: ctx.route("**/exec", relay)
    ctx.route("https://api.github.com/**", lambda x: x.abort())
    pg = ctx.new_page()
    errs = []
    pg.on("pageerror", lambda e: errs.append(str(e)))
    pg.goto(f"{base}/library/console.html"); pg.wait_for_timeout(400)
    if pg.query_selector("#thriveGate"):
        pg.fill("#gateInput", "ConThrive2030"); pg.click(".gate-btn"); pg.wait_for_timeout(1400)
    pg.evaluate(SEED)
    pg.reload(); pg.wait_for_timeout(2800)
    return ctx, pg, errs


with sync_playwright() as p:
    b = p.chromium.launch(executable_path=CH)
    ctx, pg, errs = session(b)

    st = pg.evaluate("()=>ThriveFlows.selfTest()")
    ck("the flow registry passes its own test", st["pass"], st.get("failures"))

    a = pg.evaluate("()=>ThriveFlows.audit()")
    ck("every multi-step interaction is in the registry", a["count"] >= 6, a)
    ck("and every one of them is complete", a["ok"], a["problems"])

    ck("a flow not in the registry does not open",
       pg.evaluate("()=>ThriveFlows.canOpen('invented').ok") is False)
    ck("and it says why",
       "not registered" in pg.evaluate("()=>ThriveFlows.canOpen('invented').why"))

    ck("every registered flow declares a back, a close and a completion",
       pg.evaluate("""()=>ThriveFlows.names().every(n=>{const f=ThriveFlows.get(n);
         return !!(f.back && f.close && f.completion);})"""))

    # ---- every action ends in one of exactly three outcomes ----------------
    ck("there are exactly three outcomes",
       pg.evaluate("()=>ThriveFlows.OUTCOMES.join(',')") == "success,failure,cancelled")
    ck("and a fourth throws rather than becoming a silent success",
       pg.evaluate("""()=>{try{ThriveFlows.result('maybe','x');return false;}catch(e){return true;}}"""))
    ck("every result carries a message",
       pg.evaluate("()=>!!ThriveFlows.result('failure','the relay did not answer').message"))

    # ---- every network call has a timeout and a stated failure -------------
    ck("a promise that never settles is rejected, with a message that names it",
       pg.evaluate("""async ()=>{
         try{ await ThriveFlows.withTimeout(new Promise(()=>{}), 120, 'the relay');
              return false; }
         catch(e){ return e.timeout===true && /the relay/.test(e.message)
                          && /timed out/.test(e.message); }}"""))
    ck("and a call that answers in time is untouched",
       pg.evaluate("""async ()=>{
         const v=await ThriveFlows.withTimeout(Promise.resolve('ok'), 1000, 'x'); return v==='ok';}"""))
    ck("a rejection with no message gains one",
       pg.evaluate("""async ()=>{
         try{ await ThriveFlows.withTimeout(Promise.reject(new Error('')), 1000, 'the relay'); return false; }
         catch(e){ return /the relay/.test(e.message); }}"""))

    # ---- the console's own calls are time boxed ----------------------------
    # A route that never answers, which is what a cold Apps Script deployment on a
    # weak network looks like from the browser's side.
    pg.route("**/stalls-forever", lambda r: None)
    ck("the console's relay helper aborts rather than hanging",
       pg.evaluate("""async ()=>{
         try{ await fetchT('/stalls-forever', {}, 300); return false; }
         catch(e){ return !!e.timeout || /abort|time/i.test(e.message||''); }}"""))

    # ---- the registry gates the one entry point ---------------------------
    pg.evaluate("()=>location.hash='#board'"); pg.wait_for_timeout(1300)
    pg.click(".tok[data-slug='one']"); pg.wait_for_timeout(1200)
    ck("a registered flow opens", pg.eval_on_selector("#modal", "e=>!e.hidden"))
    pg.keyboard.press("Escape"); pg.wait_for_timeout(800)
    if pg.query_selector("#threeWay"): pg.click("#threeWay [data-tw='2']"); pg.wait_for_timeout(600)

    broke = pg.evaluate("""async ()=>{
      const saved=ThriveFlows.FLOWS.opportunity.completion;
      ThriveFlows.FLOWS.opportunity.completion="";
      await window.thriveModal.open('one');
      const opened=!document.getElementById('modal').hidden;
      ThriveFlows.FLOWS.opportunity.completion=saved;
      return opened;}""")
    ck("and a flow whose completion is removed refuses to open", broke is False)
    pg.click(".tok[data-slug='one']"); pg.wait_for_timeout(1200)
    ck("and it opens again once the declaration is whole",
       pg.eval_on_selector("#modal", "e=>!e.hidden"))
    pg.keyboard.press("Escape"); pg.wait_for_timeout(800)
    if pg.query_selector("#threeWay"): pg.click("#threeWay [data-tw='2']"); pg.wait_for_timeout(600)

    ck("nothing threw", not errs, errs[:3])
    ctx.close()
    b.close()

httpd.shutdown()

# ---- the build gate, proven by breaking it --------------------------------
src = os.path.join(ROOT, "library", "flows.js")
bak = src + ".bak"
shutil.copy(src, bak)
try:
    with open(src, encoding="utf-8") as fh:
        s = fh.read()
    broken = s.replace('      back: "modalBack",\n      close: "modalClose",\n'
                       '      completion: "the message is sent and enters the mail ledger, or it is not",',
                       '      back: "",\n      close: "modalClose",\n'
                       '      completion: "the message is sent and enters the mail ledger, or it is not",', 1)
    ck("the harness could actually break the registry", broken != s)
    with open(src, "w", encoding="utf-8") as fh:
        fh.write(broken)
    r = subprocess.run(["node", os.path.join(ROOT, "tools", "verify.js")],
                       capture_output=True, text=True, cwd=ROOT)
    ck("removing a flow's back fails the build", r.returncode != 0, r.stdout[-400:])
    ck("and the build says which flow and what is missing",
       "compose: no back" in r.stdout, r.stdout[-400:])
finally:
    shutil.move(bak, src)

r = subprocess.run(["node", os.path.join(ROOT, "tools", "verify.js")],
                   capture_output=True, text=True, cwd=ROOT)
ck("and the build is green again once it is put back", r.returncode == 0, r.stdout[-300:])

print("\n%d failed" % len(fails))
for f in fails: print("  -", f)
sys.exit(1 if fails else 0)

File: library/flows.js
/* One registry, no open loops.
   =========================================================================

   A flow that can be entered and not left is a dead end, and the review found
   several. The fix that lasts is not finding them one at a time: it is making a
   flow that has no ending impossible to declare, and failing the build on one
   that is.

   EVERY MULTI-STEP INTERACTION IS DECLARED HERE, with its entry, its steps, its
   exits and its completion. A flow not in the registry does not open.

   The gate is in tools/verify.js and it reads this file. It fails on a
   registered flow with no back, no close, or no defined completion, and the
   proof that it works is removing one and watching the build go red.

   Pure. No DOM, no storage, no network. */
(function (root, factory) {
  var api = factory();
  if (typeof module === "object" && module.exports) module.exports = api;
  root.ThriveFlows = api;
})(typeof globalThis !== "undefined" ? globalThis : this, function () {
  "use strict";

  /* Every action ends in one of exactly these three, and each one says what
     happened. A fourth outcome is a spinner that runs forever. */
  var OUTCOMES = ["success", "failure", "cancelled"];

  /* Every network call has a timeout and a stated failure. The relay can be
     slow, and a promise that never settles is a frozen console. */
  var DEFAULT_TIMEOUT_MS = 30000;

  var FLOWS = {
    opportunity: {
      title: "mw_overview",
      entry: "a card on the board",
      steps: ["overview", "text", "page", "outreach", "history"],
      back: "modalBack",
      close: "modalClose",
      completion: "the window closes, and the card carries whatever moved",
      outcomes: OUTCOMES
    },
    outreach_channel: {
      title: "och_h",
      entry: "the Outreach tab of an opportunity",
      steps: ["question", "channel", "sent"],
      back: "ochChange",
      close: "modalClose",
      completion: "send_offchannel is recorded, or the path is changed",
      outcomes: OUTCOMES
    },
    compose: {
      title: "cmp_title",
      entry: "the Outreach tab, email path",
      steps: ["write", "check", "send"],
      back: "modalBack",
      close: "modalClose",
      completion: "the message is sent and enters the mail ledger, or it is not",
      outcomes: OUTCOMES
    },
    upload_kind: {
      title: "tpl_upload_h",
      entry: "the Library upload control",
      steps: ["read", "decide", "save"],
      back: "modalBack",
      close: "modalClose",
      completion: "one sentence saying what was decided and where it went",
      outcomes: OUTCOMES
    },
    intake_batch: {
      title: "in_batch_h2",
      entry: "the Library import control",
      steps: ["read", "report", "write"],
      back: "modalBack",
      close: "modalClose",
      completion: "the report is shown and the chosen records are written",
      outcomes: OUTCOMES
    },
    reply_repair: {
      title: "rp_repair_btn",
      entry: "Settings, Replies",
      steps: ["count", "confirm", "write"],
      back: "modalBack",
      close: "modalClose",
      completion: "the count is reported, then written or abandoned",
      outcomes: OUTCOMES
    }
  };

  function names() { return Object.keys(FLOWS); }
  function get(name) { return FLOWS[name] || null; }

  /* A flow not in the registry does not open. This is the enforcement, and it
     returns a reason rather than a boolean so the caller can say why. */
  function canOpen(name) {
    var f = get(name);
    if (!f) return { ok: false, why: "not registered: " + String(name) };
    var bad = problems(name);
    if (bad.length) return { ok: false, why: bad.join("; ") };
    return { ok: true };
  }

  /* What is wrong with a declaration, in words. The build gate prints these. */
  function problems(name) {
    var f = get(name), out = [];
    if (!f) return ["not registered"];
    if (!f.entry) out.push("no entry");
    if (!f.steps || !f.steps.length) out.push("no steps");
    if (!f.back) out.push("no back");
    if (!f.close) out.push("no close");
    if (!f.completion) out.push("no defined completion");
    if (!f.outcomes || OUTCOMES.some(function (o) { return f.outcomes.indexOf(o) < 0; })) {
      out.push("does not declare all three outcomes");
    }
    return out;
  }

  function audit() {
    var bad = [];
    names().forEach(function (n) {
      var p = problems(n);
      if (p.length) bad.push(n + ": " + p.join(", "));
    });
    return { ok: !bad.length, problems: bad, count: names().length };
  }

  /* ---- every action ends somewhere --------------------------------------- */

  /* One shape for the end of an action, so no caller can invent a fourth. A
     result with no outcome is the silent success this exists to prevent. */
  function result(outcome, message, detail) {
    if (OUTCOMES.indexOf(outcome) < 0) {
      throw new Error("unknown outcome: " + String(outcome));
    }
    return { outcome: outcome, message: String(message || ""), detail: detail };
  }

  /* Every network call has a timeout and a STATED failure. A rejection with no
     message is a spinner with extra steps. */
  function withTimeout(promise, ms, label) {
    ms = ms || DEFAULT_TIMEOUT_MS;
    return new Promise(function (resolve, reject) {
      var done = false;
      var timer = setTimeout(function () {
        if (done) return;
        done = true;
        var e = new Error(String(label || "request") + " timed out after " + ms + "ms");
        e.timeout = true;
        reject(e);
      }, ms);
      Promise.resolve(promise).then(function (v) {
        if (done) return;
        done = true; clearTimeout(timer); resolve(v);
      }, function (e) {
        if (done) return;
        done = true; clearTimeout(timer);
        if (e && !e.message) e = new Error(String(label || "request") + " failed");
        reject(e);
      });
    });
  }

  function selfTest() {
    var f = [];
    var a = audit();
    if (!a.ok) f.push("every registered flow must be complete: " + a.problems.join(" | "));
    if (a.count < 6) f.push("the registry is thinner than the console, got " + a.count);

    if (canOpen("opportunity").ok !== true) f.push("a complete flow must open");
    if (canOpen("nothing-like-this").ok !== false) f.push("an unregistered flow must not open");
    if (!/not registered/.test(canOpen("nope").why || "")) f.push("and it must say why");

    /* The gate, proven by breaking one. */
    var saved = FLOWS.compose.back;
    FLOWS.compose.back = "";
    if (audit().ok) f.push("a flow with no back must fail the audit");
    if (canOpen("compose").ok) f.push("and must not open");
    FLOWS.compose.back = saved;
    var saved2 = FLOWS.compose.completion;
    FLOWS.compose.completion = "";
    if (audit().ok) f.push("a flow with no completion must fail the audit");
    FLOWS.compose.completion = saved2;
    if (!audit().ok) f.push("and the registry must be clean again afterwards");

    try { result("maybe", "x"); f.push("a fourth outcome must throw"); } catch (e) {}
    var r = result("failure", "the relay did not answer");
    if (r.outcome !== "failure" || !r.message) f.push("a result must carry its message");

    return { pass: !f.length, failures: f };
  }

  return {
    OUTCOMES: OUTCOMES,
    DEFAULT_TIMEOUT_MS: DEFAULT_TIMEOUT_MS,
    FLOWS: FLOWS,
    names: names, get: get, canOpen: canOpen, problems: problems, audit: audit,
    result: result, withTimeout: withTimeout,
    selfTest: selfTest
  };
});
